src/python/pants/bsp/context.py

```diff
@@ -3,9 +3,10 @@
 from __future__ import annotations
 
 import threading
+from pathlib import Path
```

Comment:
Generally `Path` should be avoided and we use `PurePath` to make sure we don't violate the engine's assumptions on IO side effects. Is that possible here?

Reply:
The code that will use this in follow-on PRs will actually be accessing the filesystem. Those rules will have side effects; for example, writing out classfiles for Java/Scala code.

Reply:
Sounds like they won't be using `Workspace.write_digest()`? That's fine?

Reply:
I probably will have to pass `Workspace` into some of the rules to allow writing of compilation results into the tempdir. I haven't yet reached the point where I will need to do that, but that point will be the very next PR. The next PR will wire up compilation for Scala.

Reply:
Okay, this is fine because it's experimental, but something to pay attention to if we can swap this out for `PurePath`.